src/rpp/rpp/schedulers/trampoline_scheduler.hpp

```diff
@@ -42,52 +36,83 @@ struct scope_guard {
+        return s_queue;
+    }
+
+    static void drain_queue()
+    {
+        auto& queue          = get_schedulable_queue();
+        auto  reset_at_final = utils::finally_action{[] { get_schedulable_queue().reset(); }};
+
+        while (!queue->empty())
+        {
+            const auto& top = queue->top();
+
+            auto function = wait_and_extract_executable_if_subscribed(top);
+
+            // firstly we need to pop schedulable from queue due to execution of function can add new schedulable
+            queue->pop();
+
+            if (function)
+                function();
+        }
+    }
+
+    [[nodiscard]] static std::function<void()> wait_and_extract_executable_if_subscribed(const  current_thread_schedulable& schedulable)
+    {
+        if (!schedulable.is_subscribed())
+            return {};
+
+        std::this_thread::sleep_until(schedulable.GetTimePoint());
+
+        if (!schedulable.is_subscribed())
+            return {};
+
+        return std::move(schedulable.ExtractFunction());
+    }
+
 public:
     class worker_strategy
     {
     public:
         explicit worker_strategy(const rpp::composite_subscription& subscription)
-            : m_sub{subscription} {};
+            : m_sub{subscription} {}
 
         void defer_at(time_point time_point, std::invocable auto&& fn) const
         {
             if (!m_sub.is_subscribed())
                 return;
 
-            with_thread_local_schedulable_queue([&](auto& queue, bool& queue_in_use)
-            {
-                queue.emplace(time_point, std::forward<decltype(fn)>(fn));
-
-                if (queue_in_use)
-                    return;
-
-                queue_in_use = true;
-                scope_guard cleanup([&]() noexcept
-                {
-                    // If error occurs, we still want the thread-local state to be reset.
-                    queue_in_use = false;
-                    if (!m_sub.is_subscribed())
-                        queue.reset();
-                });
-
-                std::stop_source stop;
-                auto stop_token = stop.get_token();
-                m_sub.add([stop = std::move(stop)]() mutable
-                {
-                    // The following blocking-pop exits when subscription is unsubscribed in other thread.
-                    stop.request_stop();
-                });
-
-                while (!queue.is_empty() && m_sub.is_subscribed())
-                {
-                    std::function<void()> schedulable{};
-                    if (queue.pop_with_wait(schedulable, stop_token))
-                    {
-                        schedulable();
-                        schedulable = {};
-                    }
-                }
-            });
+            auto drain_handle = ensure_queue_if_no_any_owner();
+            get_schedulable_queue()->emplace(time_point, std::forward<decltype(fn)>(fn), m_sub);
```

Comment:
I understand the needs for reuse of `ensure_queue_if_no_any_owner()`. But here is a concern in terms of the readability.

The code was flat and the control flow is explained in one place. But with calling nested function here, I'm worried the readers will need stronger background knowledge for behavior of **trampoline** to read the code. For example:

1. Readers need to dive into the `ensure_queue_if_no_any_owner()` to know it's using RAII to dequeuing the message queue in the current thread.
2. Readers need to realize that we enqueue the task/schedulable before #1 really happens.

Reply:
> Readers need to dive into the `ensure_queue_if_no_any_owner()` to know it's using RAII to dequeuing the message queue in the current thread.

You may want to add extra comment in `ensure_queue_if_no_any_owner()` for the reason why executing on destructing is important.

Reply:
Another thought:

To eliminate `std::function` from my initial implementation, `with_thread_local_schedulable_queue()`, we could use `invocable` concept and feed a member function to it.

Reply:
Thanks a lot ❤️

Reply:
Can you check if it come more understandable?